libs/libc/unistd/lib_getoptvars.c

```diff
@@ -0,0 +1,153 @@
+/****************************************************************************
+ * libs/libc/unistd/lib_getoptvars.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <unistd.h>
+#include <pthread.h>
+
+#include "unistd.h"
+#include "libc.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define INVALID_KEY    CONFIG_TLS_NELEM
+#define VALID_KEY(key) ((key) >= 0 && key < CONFIG_TLS_NELEM)
+
+/* Configuration.  Currently no thread-specific data destructors. */
+
+#undef  HAVE_DESTRUCTOR
+#define HAVE_CLEANUP   1
+#define HAVE_TLSVARS   1
+
+/* Check if we have some way to avoid a memory leak */
+
+#if defined(CONFIG_DISABLE_PTHREAD) || !defined(CONFIG_PTHREAD_CLEANUP)
+#  undef HAVE_CLEANUP
+#endif
+
+/* Don't use TLS if we cannot avoid the memory leak or if TLS is disabled */
+
+#if (!defined(HAVE_DESTRUCTOR) && !defined(HAVE_CLEANUP)) || \
+    CONFIG_TLS_NELEM == 0
+#  undef HAVE_TLSVARS
+#endif
+
```

Comment:
The above options CONFIG_DISABLE_PTHREAD and CONFIG_PTHREAD_CLEANUP options are a nuisance.  What do people think about just removing this?  Pthreads are quite large and this will impact small platforms.

Another option would be to support pthread_cleanup_push() even when pthreads are disabled.. similar to how pthread specific data is enabled now.

The pthread specific data destructor is not currently implemented.  @antmerlino started this some months back (along with correct some other issues) on the pthread_user thread but we lost steam on that one.

If the destructor were implemented, it will automatically callback to a pthread data destructor when the pthread exits.

Reply:
Is disabling pthread common? I would have assumed that it is kind of required for some parts of the OS.
Would the extra space be in FLASH mostly? If so, I think the linker usually throws away unused functions.

Reply:
> 
> 
> Is disabling pthread common? I would have assumed that it is kind of required for some parts of the OS.

Yes:

```
$ find boards -name defconfig | xargs grep CONFIG_DISABLE_PTHREAD=y
boards/arm/c5471/c5471evm/configs/nettest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/dm320/ntosd-dm320/configs/nettest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/dm320/ntosd-dm320/configs/udp/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/lpc17xx_40xx/lincoln60/configs/thttpd-binfs/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/lpc17xx_40xx/lpcxpresso-lpc1768/configs/dhcpd/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/lpc17xx_40xx/lpcxpresso-lpc1768/configs/thttpd/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/lpc17xx_40xx/olimex-lpc1766stk/configs/nettest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/lpc17xx_40xx/olimex-lpc1766stk/configs/slip-httpd/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/lpc17xx_40xx/olimex-lpc1766stk/configs/thttpd-binfs/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/lpc17xx_40xx/olimex-lpc1766stk/configs/thttpd-nxflat/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/lpc17xx_40xx/zkit-arm-1769/configs/hello/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/lpc17xx_40xx/zkit-arm-1769/configs/thttpd/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/nucleo-f302r8/configs/nsh/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/nucleo-f334r8/configs/adc/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/nucleo-f334r8/configs/nsh/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/stm3220g-eval/configs/dhcpd/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/stm3220g-eval/configs/nettest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/stm3220g-eval/configs/telnetd/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/stm3240g-eval/configs/dhcpd/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/stm3240g-eval/configs/nettest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/stm3240g-eval/configs/telnetd/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/stm32f103-minimum/configs/hello/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/stm32f334-disco/configs/buckboost/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/stm32f334-disco/configs/nsh/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/stm32/stm32f334-disco/configs/powerled/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/tiva/eagle100/configs/nettest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/arm/tiva/eagle100/configs/thttpd/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/avr/at90usb/micropendous3/configs/hello/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/avr/at90usb/teensy-2.0/configs/hello/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/avr/at90usb/teensy-2.0/configs/nsh/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/avr/atmega/amber/configs/hello/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/avr/atmega/arduino-mega2560/configs/hello/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/avr/atmega/arduino-mega2560/configs/nsh/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/avr/atmega/moteino-mega/configs/hello/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/avr/atmega/moteino-mega/configs/nsh/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/hc/m9s12/demo9s12ne64/configs/ostest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/hc/m9s12/ne64badge/configs/ostest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/mips/pic32mx/mirtoo/configs/nxffs/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/misoc/lm32/misoc/configs/nsh/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/or1k/mor1kx/or1k/configs/nsh/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/renesas/sh1/us7032evb1/configs/nsh/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/renesas/sh1/us7032evb1/configs/ostest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/sim/sim/sim/configs/configdata/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/sim/sim/sim/configs/mtdpart/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/sim/sim/sim/configs/nxffs/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/sim/sim/sim/configs/spiffs/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/z80/ez80/ez80f910200kitg/configs/ostest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/z80/ez80/ez80f910200zco/configs/dhcpd/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/z80/ez80/ez80f910200zco/configs/nettest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/z80/ez80/z20x/configs/hello/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/z80/ez80/z20x/configs/w25boot/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/z80/z180/p112/configs/ostest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/z80/z8/z8encore000zco/configs/ostest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/z80/z8/z8f64200100kit/configs/ostest/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/z80/z80/z80sim/configs/nsh/defconfig:CONFIG_DISABLE_PTHREAD=y
boards/z80/z80/z80sim/configs/ostest/defconfig:CONFIG_DISABLE_PTHREAD=y
```

> Would the extra space be in FLASH mostly? If so, I think the linker usually throws away unused functions.

Yes, mostly flash.  But it is draw in a lot of the core pthread functionality just by calling the pthread initialization logic (as I recall).